### PR TITLE
On-premise ordering

### PR DESCRIPTION
**Pull request by slemrmartin**
ServiceOffering#order operation for on-premise Towers (using receptor-client)

---

[TPINVTRY-1032](https://projects.engineering.redhat.com/browse/TPINVTRY-1032)

### CODE REVIEW

File: lib/topological_inventory/ansible_tower/operations/core/ansible_tower_client.rb [hunk 1]
```diff
@@ -120,11 +120,31 @@ def verify_ssl_mode
...
+              :base_url       => full_hostname(default_endpoint),
+              :username       => authentication.try(:username),
+              :password       => authentication.try(:password),
+              :verify_ssl     => verify_ssl_mode,
+              :receptor_node  => default_endpoint.receptor_node.to_s.strip,
+              :account_number => account_number
             )
           end
+
+          def account_number
```
Comment (slemrmartin) [inline lib/topological_inventory/ansible_tower/operations/core/ansible_tower_client.rb:135]: @lindgrenj6 I need to discuss this method. 

Can I expect that forwardable_headers from catalog will contain account_number?

Comment (lindgrenj6) [reply lib/topological_inventory/ansible_tower/operations/core/ansible_tower_client.rb:135]: Yes - it will be in the x-rh-identity header.

**Comment by slemrmartin** [comment]
ready for review

File: lib/topological_inventory/ansible_tower/receptor/api.rb [hunk 2]
```diff
@@ -23,10 +27,10 @@ def class_from_type(type)
         send("#{type}_class") if respond_to?("#{type}_class")
       end
 
-      # JobTemplate v1 is for Tower version < 3 which we don't support
+      # JobTemplate v2 is for Tower version < 3 which we don't support
       # https://bugzilla.redhat.com/show_bug.cgi?id=1369842
       def job_template_class
-        @job_template_class ||= AnsibleTowerClient::JobTemplateV2
```
Comment (syncrou) [inline lib/topological_inventory/ansible_tower/receptor/api.rb:29]: What prompted this change?

Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/receptor/api.rb:29]: @syncrou I misunderstood the condition before, V2 is for older towers, V1 for >=3
https://github.com/ansible/ansible_tower_client_ruby/blob/master/lib/ansible_tower_client/api.rb#L197

File: lib/topological_inventory/ansible_tower/operations/core/ansible_tower_client.rb [hunk 3]
```diff
@@ -120,11 +120,32 @@ def verify_ssl_mode
             default_endpoint.verify_ssl ? OpenSSL::SSL::VERIFY_PEER : OpenSSL::SSL::VERIFY_NONE
           end
 
+          # TODO: Join with availability_check methods
```
Comment (lindgrenj6) [inline lib/topological_inventory/ansible_tower/operations/core/ansible_tower_client.rb:123]: :tada: I like this idea.

File: lib/topological_inventory/ansible_tower/receptor/api_object.rb [hunk 4]
```diff
@@ -28,6 +28,11 @@ def get
         raw_kafka_response(response)
       end
 
+      def post(data)
+        response = send_request(:post, endpoint, :data => data)
+        raw_kafka_response(response)
+      end
```
Comment (lindgrenj6) [inline lib/topological_inventory/ansible_tower/receptor/api_object.rb:34]: So - as far as I understand this, we're doing a blocking request for it right? That sounds good just want to make sure I'm seeing it right.

**Review by lindgrenj6 — APPROVED** [review]

File: lib/topological_inventory/ansible_tower/receptor/api_object.rb (hunk 4 above)
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/receptor/api_object.rb:34]: yes, at this moment it's like availability check

File: lib/topological_inventory/ansible_tower/operations/core/ansible_tower_client.rb (hunk 3 above)
Comment (slemrmartin) [reply lib/topological_inventory/ansible_tower/operations/core/ansible_tower_client.rb:123]: yeah, after this and targeted refresh will be merged, we should do some refactoring of this. 
Maybe we could also implement single endpoint in Sources API to return everything in one request to simplify it

**Review by lindgrenj6 — APPROVED** [review]

File: spec/receptor/api_object_spec.rb [hunk 5]
```diff
@@ -85,6 +85,32 @@
...
+  end
+
+  describe "#post" do
+    let(:post_data) { {:service_parameters => {"username" => "user1", "quest" => "Test Topology", "airspeed" => 50}, :provider_control_parameters => {}} }
+
+    context "sync" do
+      it "sends request to the default path and type" do
+        body = {:key => :value}
+        expect(subject).to(receive(:send_request)
+                             .with(:post, '/api/v2/job_templates', :data => post_data)
```
Comment (mkanoor) [inline spec/receptor/api_object_spec.rb]: @slemrmartin shouldn't this include the job_template_id?

**Review by syncrou — COMMENTED** [review]
LGTM except for Madhu's concerns. It seems the answer to his question has potential to affect the code changes.

File: spec/receptor/api_object_spec.rb (hunk 5 above)
Comment (slemrmartin) [reply spec/receptor/api_object_spec.rb]: This is testing the common `post` method, so it doesn't matter in this spec, but I'll change it.
The ordering spec exists, anyway I'll try to extend it

Comment (slemrmartin) [reply spec/receptor/api_object_spec.rb]: ~~hm, this method was intended to be generic but it seems it doesn't make much sense, that's why the launch() method is using `send_request` directly. I'll remove it.~~
EDIT: no, sorry, faster writing than thinking :laughing:

**Review by syncrou — APPROVED** [review]
LGTM 👍 

@lindgrenj6 @mkanoor any thoughts?

**Review by mkanoor — APPROVED** [review]

**Review by lindgrenj6 — APPROVED** [review]

**Pull request merged**